examples/mnist_trainer.py

```diff
@@ -0,0 +1,325 @@
+    params = lasagne.layers.get_all_params(network, trainable=True)
+    updates = lasagne.updates.nesterov_momentum(
+            loss.mean(), params, learning_rate=0.01, momentum=0.9)
+
+    # Create a loss expression for validation/testing. The crucial difference
+    # here is that we do a deterministic forward pass through the network,
+    # disabling dropout layers.
+    test_prediction = lasagne.layers.get_output(network, deterministic=True)
+    test_loss = lasagne.objectives.categorical_crossentropy(test_prediction,
+                                                            target_var)
+    # Create an expression for the number of errors in the mini-batch.
+    # This will be used for the validation score:
+    test_err_sum = T.sum(T.neq(T.argmax(test_prediction, axis=1), target_var),
+                         dtype=theano.config.floatX)
+
+    # Compile a function performing a training step on a mini-batch (by giving
+    # the updates dictionary). Return the sum of the losses over the
+    # mini-batch in a list:
+    train_fn = theano.function([input_var, target_var], [loss.sum()],
+                               updates=updates)
+
+    # Compile a second function computing the error rate (first element,
+    # as Trainer uses the first value from the validation results to
+    # determine when the network has the best validation score) and
+    # validation loss sum:
+    eval_fn = theano.function([input_var, target_var],
+                              [test_err_sum, test_loss.sum()])
+
+    # The (optional) train logging function shows the loss (the default
+    # would often be okay)
+    train_log_func = lambda train_results: 'loss={:.6f}'.format(
+        train_results[0])
+
+    # The (optional) eval logging function shows the error rate and loss
+    # (the default would be okay here too)
+    eval_log_func = lambda eval_results: 'err={:.2%}, loss={:.6f}'.format(
+        eval_results[0], eval_results[1])
+
+    # Set off the training loop:
+    # - provide the datasets train, validation and test
+    # - provide batch size
+    # - batch training function
+    # - training log function
+    # - evaluation (validation/test) function
+    # - evaluation log function
+    # - number of epochs
+    # - verbosity (report per epoch)
+    # - layer to restore; this way the network will be set back to its
+    #   state when it achieved the best validation score
+    # - random number generator to shuffle training samples (just for show
+    #   here since it would use `lasagne.random.get_rng()` anyway
+    res = lasagne.trainer.train(
+        [X_train, y_train], [X_val, y_val], [X_test, y_test], batchsize=500,
+        train_batch_func=train_fn, train_log_func=train_log_func,
+        eval_batch_func=eval_fn, eval_log_func=eval_log_func,
+        num_epochs=num_epochs, verbosity=lasagne.trainer.VERBOSITY_EPOCH,
+        layer_to_restore=network, shuffle_rng=lasagne.random.get_rng())
```

Comment:
This seems like a pretty complicated API, especially with the log functions. I'm not a big fan. It hides the for-loop, which is nice if you're happy with a lot of defaults, but in this case it ends up requiring a bunch of lambdas to be able to customise things, which is a lot less readable than just having that stuff in the body of a for loop. I think this is probably not the best API for this use case.